compiler/noirc_evaluator/src/ssa/opt/basic_conditional.rs

```diff
@@ -0,0 +1,379 @@
+use std::collections::{HashSet, VecDeque};
+
+use acvm::AcirField;
+use fxhash::FxHashMap;
+use iter_extended::vecmap;
+
+use crate::ssa::{
+    ir::{
+        basic_block::BasicBlockId,
+        cfg::ControlFlowGraph,
+        dfg::DataFlowGraph,
+        function::{Function, FunctionId, RuntimeType},
+        function_inserter::FunctionInserter,
+        instruction::{BinaryOp, Instruction, TerminatorInstruction},
+        post_order::PostOrder,
+        value::ValueId,
+    },
+    Ssa,
+};
+
+use super::flatten_cfg::Context;
+#[derive(Debug, Clone)]
+struct BasicConditional {
+    block_entry: BasicBlockId,
+    block_then: Option<BasicBlockId>,
+    block_else: Option<BasicBlockId>,
+    block_exit: BasicBlockId,
+}
+
+impl Ssa {
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub(crate) fn flatten_basic_conditionals(mut self) -> Ssa {
+        // Retrieve the 'no_predicates' attribute of the functions in a map, to avoid problems with borrowing
+        let mut no_predicates = FxHashMap::default();
+        for function in self.functions.values() {
+            no_predicates.insert(function.id(), function.is_no_predicates());
+        }
+        for function in self.functions.values_mut() {
+            flatten_function(function, &mut no_predicates);
+        }
+        self
+    }
+}
+
+// Returns the blocks of the simple conditional sub-graph whose input block is the entry.
+// Returns None if the input block is not the entry block of a simple conditional.
+fn is_conditional(
+    block: BasicBlockId,
+    cfg: &ControlFlowGraph,
+    function: &Function,
+) -> Option<BasicConditional> {
+    // jump overhead is the cost for doing the conditional and jump around the blocks
+    // I use 10 as a rough estimate, real cost is less.
+    let jump_overhead = 10;
+    let mut successors = cfg.successors(block);
+    let mut result = None;
+    // a conditional must have 2 branches
+    if successors.len() == 2 {
```

Comment:
Let's just early return with `None` if `successors.len() != 2` to reduce this nesting

Reply:
Done